portable/GCC/ARM_CRx_MPU/mpu_wrappers_v2_asm.S

```diff
@@ -0,0 +1,1333 @@
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/*---------------------------------------------------------------------------*/
+
+    .arm
+    .syntax unified
+    .section freertos_system_calls
+#define FREERTOS_ASSEMBLY
+    #include "FreeRTOSConfig.h"
+    #include "portmacro_asm.h"
+    #include "mpu_syscall_numbers.h"
+#undef FREERTOS_ASSEMBLY
+
+/* ------------------ Start of Port Specific System Calls ------------------ */
+
+/**
+ * Function: void vPortYield
+ * Inputs:      VOID
+*/
+.align 4
+.global vPortYield
+.type vPortYield, %function
+vPortYield:
+    /* Make the SVC to swap tasks */
+    SVC     #portSVC_YIELD
+    /* After yielding to another task, resume executing the calling task */
+    BX  LR
+
+/*-------------------------------------------------------------------------------*/
+/* vPortSystemCallExit */
+.align 4
+.global vPortSystemCallExit
+.type vPortSystemCallExit, %function
+vPortSystemCallExit:
+    /* Make the SVC to exit a system call */
+    SVC     #portSVC_SYSTEM_CALL_EXIT
+    /* After performing the requested FreeRTOS API, return to the calling task */
+    BX      LR
+
+/*-------------------------------------------------------------------------------*/
+/**
+ * Function: BaseType_t xPortIsPrivileged
+ * Inputs:      VOID
+*/
+.align 4
+.weak xPortIsPrivileged
+.type xPortIsPrivileged, %function
+xPortIsPrivileged:
+    /* Load value of SPSR into R0 */
+    MRS     R0, CPSR
+    /* Get the relevant mode bits */
+    AND     R0, R0, #0x1F
+    /* Check if we're in user mode */
+    CMP     R0, #USER_MODE
+    /* If we are in user mode set R0 to 0 for the return */
+    MOVEQ   R0, #0x0
+    /* Otherwise set R0 to 1 for the return */
+    MOVNE   R0, #0x01
+    BLX      LR
+
+/*-------------------------------------------------------------------------------*/
+/**
+ * Function: UBaseType_t ulPortCountLeadingZeros
+ * Inputs:      UBaseType_t ulBitmap
+*/
+.align 4
+.weak ulPortCountLeadingZeros
+.type ulPortCountLeadingZeros, %function
+ulPortCountLeadingZeros:
+    /* Count the leading zeros and return in R0 */
+    CLZ     R0, R0
+    BX      LR
+
+/* ------------------- End of Port Specific System Calls ------------------- */
+
+.macro processorInUserMode
+    /* Push R0 before using it */
+    PUSH    {R0}
+    /* Load value of SPSR into R0 */
+    MRS     R0, CPSR
```

Comment:
Typo?
Difference between comment and code

Reply:
Comment has been corrected